### PR TITLE
Migrate Tools::str2url to new code

### PR DESCRIPTION
**Pull request by FabienPapet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This migrates the Tools::str2url with proper configuration and tests. Replace usages in legacy.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | I think QA By dev as it is covered by unit tests, we can add more tests if needed

### CODE REVIEW

File: src/Core/String/StringUtils.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\String;
+
+use Transliterator;
+
+class StringUtils
```
Comment (zuk3975) [inline src/Core/String/StringUtils.php]: we have some similar class already, I think there is no need to create additional one.
https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Util/String/StringModifier.php#L32
Btw the argument $allowAccentedChars should probably go to the method itself and not the constructor.

Comment (FabienPapet) [reply src/Core/String/StringUtils.php]: > we have some similar class already, I think there is no need to create additional one.

did not know about that class TY

>  Btw the argument $allowAccentedChars should probably go to the method itself and not the constructor.

Agree, I will fix it

File: src/Core/Util/String/StringModifierInterface.php [hunk 2]
```diff
@@ -28,6 +28,8 @@
 
 /**
  * Defines reusable methods for strings modifications.
+ *
+ * @method str2url(string $string, bool $allowAccentedChars)
```
Comment (FabienPapet) [inline src/Core/Util/String/StringModifierInterface.php]: @PrestaShop/prestashop-core-developers Should I break the interface here or should I leave it that way ?

Comment (atomiix) [reply src/Core/Util/String/StringModifierInterface.php]: I would say leave it that way, maybe advert that for 9.0 it will be part of the interface?

**Review by marsaldev — DISMISSED** [review]
LGTM

File: classes/Tools.php [hunk 3]
```diff
@@ -1401,45 +1402,13 @@ public static function link_rewrite($str, $utf8_decode = null)
      */
```
Comment (marsaldev) [inline classes/Tools.php:1401]: Do you think we should add `@deprecated`?

Comment (FabienPapet) [reply classes/Tools.php:1401]: We may do this later, but for now it is mostly code factorisation, moving code from legacy to it's tested version in the core.

File: src/Core/Util/String/StringModifier.php [hunk 4]
```diff
@@ -57,4 +64,39 @@ public function cutEnd(string $string, int $expectedLength): string
...
+        // Remove all non-whitelist chars.
+        if ($allowAccentedChars) {
+            $return_str = preg_replace('/[^a-zA-Z0-9\s\'\:\/\[\]\-\p{L}]/u', '', $return_str);
+        } else {
+            if (null === $this->transliterator) {
+                $this->transliterator = Transliterator::create('Any-Latin; Latin-ASCII');
+            }
+
+            $return_str = $this->transliterator->transliterate($string);
+            $return_str = preg_replace('/[^a-zA-Z0-9\s\'\:\/\[\]\-]/', '', $return_str);
```
Comment (mflasquin) [inline src/Core/Util/String/StringModifier.php]: Why do not try to move also replaceAccentedChars function ?

Comment (FabienPapet) [reply src/Core/Util/String/StringModifier.php]: Done !

**Review by mflasquin — DISMISSED** [review]

**Review by marsaldev — DISMISSED** [review]

File: classes/Tools.php (hunk 3 above)
Comment (zuk3975) [reply classes/Tools.php:1401]: hmm I see you added a deprecation notice (since 8.1) on above method "link_rewrite" in some older PR https://github.com/PrestaShop/PrestaShop/pull/30357/files#diff-0175e8ccd89d320e7f91e6333bc749edb922cd119188d0545f51fa3dd838aa2cR1374 which points to using this str2url instead.
So I think you should update it to suggest using the new StringUtils method str2url instead (as 8.1 is not yet released it doesn't make sense forwarding one method to another which forwards to yet another :smile:). And anyway, I think its a good idea to deprecate this str2url too as you just introduced another better method, so why not getting rid of the current "bad" one in 9.0 if we can.

TLDR - I think we should deprecate both Tools::link_rewrite and Tools::str2url, get rid of them until 9.0 and suggest using your new method.

File: src/Core/Util/String/StringModifierInterface.php [hunk 5]
```diff
@@ -28,6 +28,9 @@
 
 /**
  * Defines reusable methods for strings modifications.
+ *
+ * @method str2url(string $string, bool $allowAccentedChars) will be added in 9.0
```
Comment (jolelievre) [inline src/Core/Util/String/StringModifierInterface.php]: I didn't know this notation, good to know!!

File: classes/Tools.php (hunk 3 above)
Comment (jolelievre) [reply classes/Tools.php:1401]: So there will be another PR after to replace all the current usages of `Tools::str2url` right? And that's when we should deprecate the method I guess

**Review by jolelievre — DISMISSED** [review]
Thanks @FabienPapet

File: src/Core/Util/String/StringModifierInterface.php (hunk 5 above)
Comment (marsaldev) [reply src/Core/Util/String/StringModifierInterface.php]: me neither 😊

File: classes/Tools.php (hunk 3 above)
Comment (FabienPapet) [reply classes/Tools.php:1401]: I think this is too soon to deprecate the `Tools::str2url` function. Before deprecating we would have to replace all usages by copying  the following code into all the current function usages.

```php
$str = (new StringModifier())->str2url((string) $str, (bool) Configuration::get('PS_ALLOW_ACCENTED_CHARS_URL'));
```

This make the code less readable IMO. I think we must wait for the symfony container to be available in both frontend and backend environments, so we can replace usage by just a call to the container (as there is a dependency on the Configuration). And then deprecate the call.

**Comment by FabienPapet** [comment]
@jolelievre yes there will be a PR. (could not reply on your comment)

File: classes/Tools.php (hunk 3 above)
Comment (zuk3975) [reply classes/Tools.php:1401]: (`frontend and backend environments,` - you probably mean front office and back office :smile: ) anyway, okay, at least there is a plan :+1:

**Review by nicosomb — DISMISSED** [review]

**Comment by HanaRebaiQA** [comment]
Hello @FabienPapet 

I have launched the automated tests on this PR. 
I have failed tests on Orders and Shop-parameters https://github.com/HanaRebaiQA/testing_pr/actions/runs/3591363867/jobs/6046945669 

I checked on the nightly report of develop branch of 01/12 and it's OK
ping @PrestaShop/qa-automation : Could you please check it ?

Thank You!

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @FabienPapet 

I have launched the automated tests on this PR. 
I have failed tests on Orders and Shop-parameters https://github.com/HanaRebaiQA/testing_pr/actions/runs/3591363867/jobs/6046945669 

I checked on the nightly report of develop branch of 01/12 and it's OK
ping @PrestaShop/qa-automation : Could you please check it ?

Thank You!

**Comment by jolelievre** [comment]
Hi @FabienPapet 

can you check this use case https://github.com/HanaRebaiQA/testing_pr/actions/runs/3591363867/jobs/6046945669
Seems like it's related to URLs separator of attribute anchor so it has chance to be related to your modification

**Review by marsaldev — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Comment by paulnoelcholot** [comment]
Hello @FabienPapet !
I have tested this PR and the auto test is NOK as you can see : 

image

**Comment by FabienPapet** [comment]
Automated test result : https://github.com/FabienPapet/ga.tests.ui.pr/actions/runs/3632465683

**Comment by FabienPapet** [comment]
test auto ✅

**Review by zuk3975 — APPROVED** [review]

File: src/Core/Util/String/StringModifier.php [hunk 6]
```diff
@@ -57,4 +64,40 @@ public function cutEnd(string $string, int $expectedLength): string
...
+     * Replace all accented chars by their equivalent non-accented chars.
+     *
+     * @param string $string
+     *
+     * @return string
+     */
+    public function replaceAccentedChars(string $string): string
+    {
+        if (null === $this->transliterator) {
+            $this->transliterator = Transliterator::create('Any-Latin; Latin-ASCII');
```
Comment (zuk3975) [inline src/Core/Util/String/StringModifier.php:98]: I've already approved, but Im not sure if there is a point having transliterator here as a class property? I think it could be replaced by local variable instead.
Different transliterators might need to be used in different methods anyway.

Comment (FabienPapet) [reply src/Core/Util/String/StringModifier.php:98]: I kept the same behaviour as the previous code, and it also avoids to create an object every time you call this function.

To be honest I don't like this here (why transliterate to latin-ascii by default ???), but it is the way it works now and is better than the legacy code we have, and it's tested at least, so refactoring won't be difficult.

**Comment by FabienPapet** [comment]
@kpodemski @marsaldev I need your approval again please 🙏

**Review by kpodemski — APPROVED** [review]

**Comment by zuk3975** [comment]
automated test running here https://github.com/zuk3975/testing_pr/actions/runs/3639039911

**Review by MhiriFaten — APPROVED** [review]
Hello  @FabienPapet ,

The automated tests are OK https://github.com/djoelleuch/ga.tests.ui.pr/actions/runs/3646800842/jobs/6159244604

It's QA aprouved :heavy_check_mark: 

Thank you !

**Comment by kpodemski** [comment]
Thanks @FabienPapet
